Blog post on enhanced preview functionality (#4479)
pulumi/docs: content/blog/announcing-pulumi-preview-improvments/index.md

---
title: "Announcing Pulumi Preview Improvements"
date: 2020-10-27
draft: false
meta_desc: "Announcing a significant improvement to the pulumi preview experience."
#meta_image: 
authors:
    - paul-stack
tags:
    - enhancement
    - pulumi
---

Today we are announcing a fairly significant improvement to the Pulumi preview experience when working with Pulumi applications.
Previously, the experience may have been misleading in that some dependent resources would be seen to needing to a "replace" operation as
reported during the preview, but when apply ran, the "replace" operation didn't actually need to happen.

<!--more-->

## Change in Behaviour

We can look at a sample Pulumi application that creates a Google GKE cluster. Pulumi then deploys an application into the
kubernetes cluster.

> ```typescript

Reviewer: {{< gist pulumipus 0c6e4ede557600e114885869ea0aeb96 >}}

Please delete the rest of the code block because it's just a wall of code that a reader has to scroll through. I created a github account for Pulumipus where we can post gists and then use Hugo shortcodes to embed them. They advantage is that they only display 10 or so lines of code and has a widget that lets you expand the window if you want to see all the code.

Author: Amazing!

const deployment = new k8s.apps.v1.Deployment(name,
    {
        metadata: {
            namespace: namespaceName,
            labels: appLabels,
        },
        spec: {
            replicas: 1,
            selector: { matchLabels: appLabels },
            template: {
                metadata: {
                    labels: appLabels,
                },
                spec: {
                    containers: [
                        {
                            name: name,
                            image: "nginx:latest",
                            ports: [{ name: "http", containerPort: 80 }],
                        },
                    ],
                },
            },
        },
    },
    {
        provider: clusterProvider,
    },
);

// Create a LoadBalancer Service for the NGINX Deployment
const service = new k8s.core.v1.Service(name,
    {
        metadata: {
            labels: appLabels,
            namespace: namespaceName,
        },
        spec: {
            type: "LoadBalancer",
            ports: [{ port: 80, targetPort: "http" }],
            selector: appLabels,
        },
    },
    {
        provider: clusterProvider,
    },
);
```

> A simple change to the labels for the GKE cluster led to a misleading preview experience that told users that the kubernetes

Reviewer: ```suggestion
A simple change to the labels for the GKE cluster led to a misleading preview experience that told users that the Kubernetes
```

resources were going to be replaced:

```
     Type                              Name                         Plan        Info
     pulumi:pulumi:Stack               gcp-ts-gke-hello-world-dev2
 ~   ├─ gcp:container:Cluster          helloworld                   update      [diff: ~resourceLabels]
 +-  ├─ gcp:container:NodePool         primary-node-pool            replace     [diff: ~location]
 +-  ├─ pulumi:providers:kubernetes    helloworld                   replace     [diff: ~kubeconfig]
 +-  ├─ kubernetes:core/v1:Namespace   helloworld                   replace     [diff: -metadata~provider]
 +-  ├─ kubernetes:core/v1:Service     helloworld                   replace     [diff: ~metadata,provider]
 +-  └─ kubernetes:apps/v1:Deployment  helloworld                   replace     [diff: ~metadata,provider]

Resources:
    ~ 1 to update
    +-5 to replace
    6 changes. 1 unchanged
```

> The result of the Pulumi apply showed that only the `gcp:container:Cluster` changed and no replacements to the Kubernetes resources occurred.

Reviewer: ```suggestion
The result of the Pulumi apply showed that only the gcp:container:Cluster changed, and no other Kubernetes resources were replaced.
```

```
     Type                      Name                         Status      Info
     pulumi:pulumi:Stack       gcp-ts-gke-hello-world-dev2
 ~   └─ gcp:container:Cluster  helloworld                   updated     [diff: ~resourceLabels]

Resources:
    ~ 1 updated
    6 unchanged
```

> This was much more inline with what we expected to happen. The new preview functionality ensures an output much more aligned to what will actually happen:

Reviewer: ```suggestion
This change was what we expected to happen. The new preview functionality ensures a correct output on preview:
```

```
     Type                      Name                         Plan       Info
     pulumi:pulumi:Stack       gcp-ts-gke-hello-world-dev2
 ~   └─ gcp:container:Cluster  helloworld                   update     [diff: ~resourceLabels]

Resources:
    ~ 1 to update
    6 unchanged
```

## Conclusion

> This new functionality is available to use now as part of Pulumi v2.12.0 and is supported in all new provider releases.

Reviewer: ```suggestion
The improved preview is available in Pulumi v2.12.0 and is supported in all new provider releases.
```

---

> title: "Announcing Pulumi Preview Improvements"

Reviewer 2: Maybe skip "announcing" since this is more tactical?  Would probably drop this word throughout.

```suggestion
title: "Improved Pulumi Previews"
```

Author: Ageed!

date: 2020-10-27
draft: false
meta_desc: "Announcing a significant improvement to the pulumi preview experience."
#meta_image: 
authors:
    - paul-stack
tags:
    - enhancement
    - pulumi
---

Today we are announcing a minor but significant improvement to the Pulumi [preview]({{< relref "/docs/reference/cli/pulumi_preview" >}}
experience.

<!--more-->

## Change in Behaviour

We can look at a sample Pulumi application that creates a Google GKE cluster. Pulumi then deploys an application into the
Kubernetes cluster.

{{< gist pulumipus 0c6e4ede557600e114885869ea0aeb96 >}}

A simple change to the labels for the GKE cluster led to a misleading preview experience that told users that the Kubernetes

> resources were going to be replaced:

Reviewer 2: We should try to frame this differently.  Explain why this previous behaviour was the way it was, not frame it as "bad", but frame the opportunity to make it better.  

Something like:
1. Here's what you previously saw during the preview.
2. Note that the actual update would not need to replace any of those resources
3. This is because Pulumi was trying to be conservatively correct - let you know everything that *could* happen
4. It thought that the update of the Cluster could potentially lead to outputs of the cluster changing which other resources depend on.
5. In practice though, that won't ever happen here.  But Pulumi previously didn't have a way to check for that, and so had to be conservative to be correct.

And then note later what changed.
1. Pulumi now supports asking the individual resource providers about what outputs will change as part of an update to a resource.  
2. This allows the providers (like the GCP provider in this case) to give more specific information, which allows Pulumi to provide more accurate previews.

(Would be good for @pgavlin to review what I wrote there - but I think if we are going to do a post here we likely want a narrative along those lines).

Author: I really like this - this sums it up way nicer - thank you!

Reviewer 3: > Pulumi now supports asking the individual resource providers about what outputs will change as part of an update to a resource.

"how outputs will change" more accurately describes what the changes do, as we also include information about specific values.

>  This allows the providers (like the GCP provider in this case) to give more specific information, which allows Pulumi to provide more accurate previews.

👍

Reviewer 3: The description of our prior behavior is spot on.

 +-  ├─ kubernetes:core/v1:Service     helloworld                   replace     [diff: ~metadata,provider]
 +-  └─ kubernetes:apps/v1:Deployment  helloworld                   replace     [diff: ~metadata,provider]

Resources:
    ~ 1 to update
    +-5 to replace
    6 changes. 1 unchanged
```

The result of the Pulumi apply showed that only the `gcp:container:Cluster` changed, and no other Kubernetes resources were replaced.

```
     Type                      Name                         Status      Info
     pulumi:pulumi:Stack       gcp-ts-gke-hello-world-dev2
 ~   └─ gcp:container:Cluster  helloworld                   updated     [diff: ~resourceLabels]

Resources:
    ~ 1 updated
    6 unchanged
```

This change was what we expected to happen. The new preview functionality ensures a correct output on preview:

```
     Type                      Name                         Plan       Info
     pulumi:pulumi:Stack       gcp-ts-gke-hello-world-dev2
 ~   └─ gcp:container:Cluster  helloworld                   update     [diff: ~resourceLabels]

Resources:
    ~ 1 to update
    6 unchanged
```

> 

Reviewer 2: Possibly mention the flag to turn this off?  (or link the changlog entry?)

Author: Added the details on how to turn it off

> ## Conclusion

Reviewer 2: Possibly worth also mentioning the scenario in https://github.com/pulumi/pulumi-azure/issues/192?
